core/app/models/spree/calculator/percent_per_item.rb

```diff
@@ -0,0 +1,52 @@
+require_dependency 'spree/calculator'
+
+module Spree
+
+  # A calculator for promotions that calculates a percent-off discount
+  # for all matching products in an order. This should not be used as a
+  # shipping calculator since it would be the same thing as a flat percent
+  # off the entire order.
+  #
+  #
+  # TODO Should be deprecated now that we have adjustments at the line item level in spree core
+
+  class Calculator::PercentPerItem < Calculator
+    preference :percent, :decimal, default: 0
+
+    def self.description
+      Spree.t(:percent_per_item)
+    end
+
+    def compute(object=nil)
+      return 0 if object.nil?
+      object.line_items.reduce(0) do |sum, line_item|
+        sum += value_for_line_item(line_item)
+      end
+    end
+
+  private
+
+    # Returns all products that match this calculator, but only if the calculator
+    # is attached to a promotion. If attached to a ShippingMethod, nil is returned.
+    # Copied from per_item.rb
+    def matching_products
+      if compute_on_promotion?
+        self.calculable.promotion.rules.map do |rule|
```

Comment:
What if we cache this as `@matching_products ||= self.calculable...` since it is called multiple times in the `compute` method potentially.